determined-ai/determined
# fix: tensorboard deletion & checkpoint gc when det e delete [DET-9844]

## Description
Fix a bug where tensorboards were not deleted, even though experiments were. 

Since experiment deletion includes checkpoint garbage collection (which then calls tensorboard deletion), a consequence of this tensorboard fix is that _checkpoints are now properly garbage-collected when experiments are deleted_.

Originally, when experiments were deleted, checkpoints were deleted entirely before patched. Since the files simply didn't exist anymore, they couldn't be patched, so this crashed the garbage collection before `delete_tensorboards` could be called. As a solution, whenever experiments are deleted, do not attempt to patch the deleted checkpoints or experiment itself. 

## Test Plan
Follow these steps to create a tensorboard & delete its files.
- Spin up a brand-new devcluster with these changes.
- Edit examples/tutorials/mnist_pytorch/const.yaml (or whatever other .yaml file you'll use to create an experiment) to include: 
```
min_checkpoint_period:
     batches: 1
```
- Then, prepare to create an experiment & tensorboard:
```
$ det e create examples/tutorials/mnist_pytorch/const.yaml examples/tutorials/mnist_pytorch/
$ det tensorboard start <exp-id>
```
- Once tensor board starts, check that its files exist & are being written to `/tmp/determined-cp/`. From the root of your determined directory:
```
$ cd /tmp/determined-cp
$ find . -name 'tensorboard*' -type d
>> <folder-name>/tensorboard
$ cd <folder-name>/tensorboard/experiment/<exp-id>
// poke around in here & make sure the tensorboard files are written under the correct experiment & trial id.
$ ls 
$ cd /trial/<exp-id>
$ ls
>> events.out.tfevents.1695349683.c45fa0c3e22d.44.69	events.out.tfevents.1695349812.c45fa0c3e22d.44.140	events.out.tfevents.1695349942.c45fa0c3e22d.44.211	 ...
```
Then, if the experiment hasn't completed already (can check with `det e list`), kill it:
```
det e kill <exp-id>
```
Now delete the experiment:
```
det e delete <exp-id>
```
Check that the tensorboard files do not exist anymore:
```
$ cd /tmp/determined-cp/<folder-name>/tensorboard/experiment/
Now, you shouldn't see an <exp-id> directory when you ls:
$ ls
```
That's it! Deleting the experiment deleted the tensorboard!
- To double-check, make sure that devcluster logs include the following messages around time of experiment deletion:
```
DEBU[...] resources state changed: &{ResourcesID:ac61f4eb-54ca-4529-85fa-d6502016a99f ResourcesState:TERMINATED ResourcesStarted:<nil> ResourcesStopped:resources exited successfully with a zero exit code Container:0x14000fc1bc0}  task-id=<task-id> task-type=CHECKPOINT_GC
INFO[...] resources terminated ac61f4eb-54ca-4529-85fa-d6502016a99f: resources exited successfully with a zero exit code  task-id=<task-id> task-type=CHECKPOINT_GC
...
INFO[...] allocation stopped early after all resources exited: resources exited successfully with a zero exit code  task-id=<task-id> task-type=CHECKPOINT_GC
INFO[...] all resources are released for <task-id>.1  actor-local-addr=default actor-system=master go-type=resourcePool resource-pool=default
```
- To triple check, take the `<task-id>` of the checkpoint_gc (found in devcluster logs), run `det task logs <task-id>` to check that those logs include the appropriate delete_tensorboard logs.

## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.

## Ticket
DET-9844
